clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java

```diff
@@ -369,6 +390,78 @@ public void logUnused() {
         return objects;
     }
 
+    /**
+     * Instantiates given list of config providers and fetches the actual values of config variables from the config providers.
+     * returns a map of config key and resolved values.
+     * @param indirectVariables The map of config variables
+     * @param configProviders The map of config provider configs
+     * @param original The map of raw configs.
+     * @return map of resolved config variable.
+     */
+    @SuppressWarnings("unchecked")
+    private  Map<String, ?> resolveConfigVariables(Map<String, String> indirectVariables, Map<?, ?> configProviders, Map<String, Object> original) {
+        Map<String, ConfigProvider> providers;
+
+        if (configProviders == null || configProviders.isEmpty()) {
+            providers = instantiateConfigProvider(indirectVariables, original);
+        } else {
+
+            for (Map.Entry<?, ?> entry : configProviders.entrySet()) {
+                if (!(entry.getKey() instanceof String))
+                    configProviders.remove(entry.getKey());
+            }
+            providers = instantiateConfigProvider((Map<String, String>) configProviders, original);
+        }
+        if (!providers.isEmpty()) {
+            ConfigTransformer configTransformer = new ConfigTransformer(providers);
+            ConfigTransformerResult result = configTransformer.transform(indirectVariables);
+            original.putAll(result.data());
+        }
+
+        return original;
+    }
+
+    private Map<String, Object> getClassConfigProperties(String prefix, Map<String, ?> originals) {
+        Map<String, Object> result = new HashMap<>();
+        for (Map.Entry<String, ?> entry : originals.entrySet()) {
+            if (entry.getKey().startsWith(prefix) && entry.getKey().length() > prefix.length()) {
+                result.put(entry.getKey().substring(prefix.length()), entry.getValue());
+            }
+        }
+        return result;
+    }
+
+    private Map<String, ConfigProvider> instantiateConfigProvider(Map<String, String> indirectConfigs, Map<String, Object> originals) {
```

Comment:
`instantiateConfigProviders` since this is potentially creating multiple providers